doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+
+```
+pub async fn monotonic_appender(&self, id: GlobalId) -> CollectionAppender;
+```
+
+This new method would return a struct called `CollectionAppender`, which essentially acts as a
+`oneshot::Sender<Vec<(Row, Diff)>>`. It's a `Send`-able and non-`Clone`-able struct that has one
+method `async fn append(self, ...)` where it takes ownership of `self` so it can only be used once.
+This would be implemented as part of [`collection_mgmt`](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs)
+and would approximately be structed like:
+
+```
+struct CollectionAppender {
+  id: GlobalId,
+  tx: mpsc::Sender<(GlobalId, Vec<(Row, Diff)>)>,
+}
+```
+
+In addition to the `CollectionAppender`, we'll update the `CollectionManager` to receive a
+`oneshot::Sender<Result<(), StorageError>>` that it can use to optionally notify any listeners when
+and if the updates were successfully committed.
+
+Adding the `CollectionAppender` introduces a possible correctness issue though:
+
+1. Create a `CollectionAppender` for `GlobalId(A)`.
+2. Drop source associated with `GlobalId(A)`.
+3. Try to append data.
+
+To protect against this we can change the `CollectionManager` to return an error via the
+`oneshot::Sender` if the [provided GlobalId no longer exists](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L83).
+
+Adding this new API gets us two important properties that the existing `append` API does not have:
+
+1. The ability to send updates to the `StorageController` from a thread other than the main
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. In general webhooks are validated by
+HMAC-ing the request body using SHA256, and validating the result with a signature provided in the
+request headers. An issue though is everyone does this just a little bit different:
+
+* Segment signature = HMAC of just the body of the request
+* Buildkite signature = HMAC of "#{timestamp}.#{body}"
+* GitHub signature = "body=" + HMAC of {body}
+
+As such we'll need to support some custom logic for validation. What we can do is support a
+`VALIDATE USING` statement that accepts only a single scalar expression. This expression will
+be provided the request `headers` as `map[text] -> text`, and the `body` as `text` _regardless of
+whether or not the source has `INCLUDE HEADERS` or what `FORMAT BODY` is specified._ For example:
+
+```
+VALIDATE USING (
+  headers['X-Signature'] = hmac('sha256', SECRET webhook_secret, body)
```

Comment:
I am again concerned about the capitalization of the headers. Does the rust crate we use canonicalize them somehow?

Reply:
It looks like canonically we lowercase them? https://docs.rs/http/0.2.8/http/header/struct.HeaderName.html

Reply:
Great call outs! Yes the `http` crate always represents header names as lowercase, and when creating a `HeaderName` it will normalize it to lower case